### DISCUSSION TITLE
Afinal, quais as limitações de um BaaS com relação a código customizado?

### DISCUSSION THREAD

**Discussion by StanleySathler** [frontendbr/forum#512] [Discussão]
Olha eu de novo, meu povo! Como diz nosso amigo @woliveiras, "senta que lá vem textão".

Dessa vez, quero abrir uma discussão sobre BaaS. Calma: prometo que não vim saber sobre os "prós e contras" do _Backend as a Service_. Afinal, são 847 mil resultados do Google com esse resultado (é verdade, chequei todas as páginas :b).

_"Peraí, Backend? Isso aqui é um fórum pra Frontend, manézão! Sai daqui!"_
Bom, tô presumindo que o BaaS, por ser uma "solução rápida", seja mais frequente na comunidade dos _frontenders_. Me perdoem se eu estiver vacilando, hahaha.

**Contextualizando meu produto**

A minha questão é: estou construindo meu primeiro produto. É um aplicativo pra academias, nada viral. Mas não é só um aplicativo: tem um painel Web também, onde algumas informações são cadastradas. O aplicativo então consome essas informações. 

A academia compra, geramos uma versão pra ela para que os alunos possam utilizar. Cada academia deve ter aí, mais ou menos, uns 100, 150 alunos que seriam potenciais usuários. Por enquanto só tenho o primeiro cliente, e conseguir novos deve levar alguns meses.

Portanto, pensei em começar com um BaaS. O BaaS me garante uma API rápida, um banco de dados e um servidor. Se é a melhor opção? Bom, até então estou apostando na prática e na experiência com o serviço pra colher esse feedback (embora suas opiniões acrescentem bastante nessa dúvida).

É válido lembrar que também já desenvolvi uma puta parte do backend. É, é porque eu só conheci sobre BaaS ontem. Sendo assim, com PHP (pela Orientação a Objeto), cheguei a construir o esqueleto da API. A parte chata já está pronta, com os endpoints funcionando, chamando Controller certinho. O ORM com Doctrine também está filé. Escolhi o MySQL para me acompanhar na jornada.

**Se já fiz quase o backend inteiro, por que mudar? Qual o problema?**

Meu medo com um backend próprio é ter que ficar me preocupando com segurança dos dados (apesar de eles não serem tão sensíveis), me preocupar com o banco em si, ou com a performance do servidor. Além de que BaaS oferece algumas funcionalidades como autenticação, já que o sistema tem login. Além do mais, se vier a ser necessário no decorrer do crescimento, posso acabar investindo em um backend próprio, customizado.

Mas em contrapartida, fico pensando se lá na frente não posso precisar ter que fazer algumas peripécias no backend, como validar alguma informação ou sei lá. Qualquer coisa que possa ter que fazer no backend. Não quero ter que abandonar o BaaS e escrever um backend customizado só pra ter que adicionar uma linha de código verificando alguma coisa.

O que não ficou muito claro com as discussões que encontrei no Google foi: quais as limitações com relação a código customizado? Sei que não posso reinventar toda a roda lá dentro, mas eu posso ter *algum* código customizado? Ou absolutamente nada, basicamente ler coleções/tabelas pré-determinadas e pronto?

Ah, opiniões acerca da minha escolha pelo BaaS também são bem-vindas, embora o foco seja realmente na pergunta sobre código customizado.

Valeu, pessoal! <3

**Comment by ericdouglas** [comment]
Bem interessante sua dúvida @StanleySathler , tenho interesse em saber mais sobre esse tema também, vamos ver o que a galera tem a dizer :smile:  :popcorn: 

Você pode abranger suas pesquisas buscando o termo "serverless". A Amazon tem alguns produtos nessa linha, onde basicamente você programa uma função para rodar em um endpoint e eles te cobram X por cada vez que essa função rodar. [Essa talk](https://www.youtube.com/watch?v=2oNovfw3V08&t=6s) vai te dar uma noção melhor do tema.

Um livro que você pode pelo menos ler o sumário e a intro é [esse aqui](https://pragprog.com/book/brapps/serverless-single-page-apps).

Como BaaS o principal serviço que me vem a mente é o [Firebase](https://firebase.google.com/). Se eu fosse fazer algo do tipo que você descreveu provavelmente iria usar ele.

Caso você precise de fazer algum procedimento antes de salvar o dado no BaaS, poderia usar algo do tipo [AWS Lambda](https://aws.amazon.com/pt/lambda/).

O pouco ~que acho~ que sei sobre o tema é isso :D

**Comment by StanleySathler** [comment]
Opa, e aí @ericdouglas. Foi mal a demora, haha.

Ainda não consegui assistir à talk, mas o farei. O mesmo com o livro. E quanto ao Firebase, pois é, eu estava exatamente lendo sobre ele ontem, porque queria dar uma olhada nas opções. Acho que consigo começar com a versão gratuita para o meu primeiro cliente, o que me permite ir me acostumando com o serviço e entendendo melhor como um BaaS funciona. Quando surgir o segundo, começo a comprar de verdade.

Fiquei lendo sobre o AWS Lambda e achei do caramba! Gostei muito das funcionalidades, praticamente tudo ali escrito é novo pra mim. 

Também tô no aguardo do restante do pessoal dar uma palinha sobre o assunto.

Valeu pelo help, cara! Ajudou bastante.

**Comment by ericdouglas** [comment]
@StanleySathler dá uma olhada [nesse material](http://serverless-stack.com/), achei bem interessante!

**Comment by abalad** [comment]
https://github.com/frontendbr/forum/issues/468
